### PR TITLE
Parse struct fields

### PR DESCRIPTION
**Pull request by binji**
This allows the following field formats:

* `(struct (field $name i32))`
* `(struct (field $name (mut i32)))`
* `(struct (field i32))`
* `(struct (field (mut i32)))`
* `(struct (mut i32))`
* `(struct i32)`

### CODE REVIEW

**Comment by binji** [comment]
@jgravelle-google ping

**Comment by binji** [comment]
Looks like the travis failures are flakes?

File: src/binary-reader.h [hunk 1]
```diff
@@ -51,6 +51,12 @@ struct ReadBinaryOptions {
   bool fail_on_custom_section_error = true;
 };
 
+// TODO: Move somewhere else?
```
Comment (jgravelle-google) [inline src/binary-reader.h]: Probably yes, though I couldn't say where. Seems reasonable to leave it here at least.

File: src/shared-validator.h [hunk 2]
```diff
@@ -169,10 +173,23 @@ class SharedValidator {
...
   struct FuncType {
+    FuncType() = default;
+    FuncType(const TypeVector& params, const TypeVector& results)
+        : params(params), results(results) {}
+
     TypeVector params;
     TypeVector results;
   };
 
+  using TypeMutVector = std::vector<TypeMut>;
```
Comment (jgravelle-google) [inline src/shared-validator.h]: Nit: this seems dubiously-useful, and we still have two `std::vector<TypeMut>`s in this PR. I'd either move this to be next to `TypeMut`'s declaration so it's usable everywhere, or just use naked std::vectors.

Comment (binji) [reply src/shared-validator.h]: Good point, I moved it to where `TypeMut` is defined, and used it everywhere. I like it a bit better than using `std::vector` everywhere.

**Pull request merged**